### PR TITLE
[react] Make component constructor props non-optional

### PR DESCRIPTION
**Pull request by pelotom**
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

This PR makes `props` a non-optional argument to the constructor of component classes. Until TypeScript 2.6 it hasn't mattered much, but with the advent of `--strictFunctionTypes`, the props being optional means that implementations _must_ treat them as being possibly undefined in the constructor, or else it is not considered a `React.ComponentClass` (and hence not a `React.ComponentType`). For example:

```ts
interface Props {
  x: number
}

class Foo extends React.Component<Props> {
  constructor(props: Props) {
    super(props)
    // ...
  }
}

function doSomething(C: React.ComponentType<Props>) {
  // ...
}

doSomething(Foo)
//          ^^^
// Argument of type 'typeof Foo' is not assignable to parameter of type 'ComponentType<Props>'.
//   Type 'typeof Foo' is not assignable to type 'StatelessComponent<Props>'.
//     Type 'typeof Foo' provides no match for the signature '(props: Props & { children?: ReactNode; }, context?: any): ReactElement<any> | null'.
```

Currently one has to fix this by marking the props optional:

```ts
  constructor(props?: Props) {
    super(props)
    // ... now one has to assert non-nullness (`props!`) to access props, or else use `this.props`
  }
```

But will the `props` ever actually be `undefined` there in practice? Empirically it seems like not, if `Props` itself doesn't include the possibility.

### CODE REVIEW

**Comment by pelotom** [comment]
Looks like this change impacts several downstream libraries, mostly because of the tendancy to want to call `super()` without arguments. [The docs](https://reactjs.org/docs/state-and-lifecycle.html#adding-local-state-to-a-class) however are clear that

> Class components should always call the base constructor with props.

so this seems like the correct typing. And once `--strictFunctionTypes` is in widespread use, it will be the less painful option than having to always assert non-nullness on props in the constructor.

**Comment by RyanCavanaugh** [comment]
This PR has been open and unchanged 5 days without signoff or complaint. This will be merged by a maintainer soon if there are no objections.

**Comment by pelotom** [comment]
@RyanCavanaugh can we go ahead with the merge then?

**Review by Igorbek — APPROVED** [review]

**Comment by Igorbek** [comment]
It seems that with stricter covariance `--strictFunctionTypes`, this change is required to properly accept components that properly follows the rules. For example, it was recently reported in `styled-components` repo, see https://github.com/styled-components/styled-components/issues/1287#issuecomment-341997404

File: types/jasmine-enzyme/jasmine-enzyme-tests.tsx
Comment (DovydasNavickas) [inline types/jasmine-enzyme/jasmine-enzyme-tests.tsx]: I know it's a test, but passing an empty object is not really correct here and later on - people look at tests and learn from them. I'd rather you take props in parameters and pass the object further on.

File: types/react-native-material-kit/react-native-material-kit-tests.tsx
Comment (DovydasNavickas) [inline types/react-native-material-kit/react-native-material-kit-tests.tsx]: Same here.

File: types/react-lazyload/react-lazyload-tests.tsx
Comment (DovydasNavickas) [inline types/react-lazyload/react-lazyload-tests.tsx]: Same here.

File: types/react-datepicker/react-datepicker-tests.tsx
Comment (DovydasNavickas) [inline types/react-datepicker/react-datepicker-tests.tsx]: Same here.

File: types/material-ui-pagination/material-ui-pagination-tests.tsx
Comment (DovydasNavickas) [inline types/material-ui-pagination/material-ui-pagination-tests.tsx]: Same here.

File: types/material-ui/material-ui-tests.tsx
Comment (DovydasNavickas) [inline types/material-ui/material-ui-tests.tsx]: Same here.

**Review by DovydasNavickas — CHANGES_REQUESTED** [review]

File: types/draft-js/draft-js-tests.tsx
Comment (DovydasNavickas) [inline types/draft-js/draft-js-tests.tsx]: Same here.

**Review by DovydasNavickas — CHANGES_REQUESTED** [review]

**Comment by DovydasNavickas** [comment]
I've left multiple comments, but the bottom line is:
>I know it's a test, but passing an empty object is not really correct here and later on - people look at tests and learn from them. I'd rather you take props in parameters and pass the object further on.

If tests are already being changed, I'd rather it's done properly.

File: types/jasmine-enzyme/jasmine-enzyme-tests.tsx
Comment (pelotom) [reply types/jasmine-enzyme/jasmine-enzyme-tests.tsx]: It’s perfectly correct. The type of the props is `{}`; if it wasn’t correct it wouldn’t type check.

Comment (DovydasNavickas) [reply types/jasmine-enzyme/jasmine-enzyme-tests.tsx]: With `--strictFunctionTypes` - it's correct. But `{}` used to refer to any object, didn't it? Therefore, it's not *perfectly* correct 😄

**Comment by pelotom** [comment]
@DovydasNavickas all of these examples are perfectly correct. In all of these cases type of the component props is `{}` or `null`, so you can just pass the corresponding literal to the super constructor. It would not type check if it wasn’t correct.

**Comment by DovydasNavickas** [comment]
I guess I'm trying to fix things where they're not broken 🙂

**Review by DovydasNavickas — APPROVED** [review]

**Pull request merged**
